### ISSUE TITLE
incorrect behaviour: `docker run busybox`

### ISSUE THREAD

**Issue by hickeng** [vmware/vic#1687]
Running `docker run busybox`

Regular docker: exits immediately - presumably because stdin is closed and `ash` will exit in that case

VIC: doesn't exit, but also doesn't allow interaction - presumably we're not closing stdin

**Comment by hickeng** [comment]
assigned to @sflxn for initial investigation - this could either be in the personality side or in the tether.

**Comment by sflxn** [comment]
So, initially, I thought there was something special going on at the shell process (and there may still be something there), but after doing some inspection, I noticed that the container config for the container is different between VIC and regular docker.

For VIC, I got
```
        "Config": {
            "Hostname": "df21361b9f98",
            "Domainname": "",
            "User": "",
            "AttachStdin": true,
            "AttachStdout": true,
            "AttachStderr": true,
            "Tty": false,
            "OpenStdin": true,
            "StdinOnce": true,
```

For regular Docker, I got,

```
        "Config": {
            "Hostname": "55cd1f8f6e5b",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
```

This may partially explain why we don't exit immediately.  VIC is going to attempt attaching and sh thinks there is stdin, stdout, stderr.  This is a theory.

**Comment by sflxn** [comment]
I did more investigation. The path I was looking at (inspect data) was a red herring.  I ran a debug build of docker and see that it is also doing an attach.  The container's exit causes the attach to end.  The following is the output from a debug build of the docker daemon:

```
DEBU[0014] Calling POST /v1.24/containers/565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18/attach?stderr=1&stdout=1&stream=1 
DEBU[0014] attach: stderr: begin                        
DEBU[0014] attach: stdout: begin                        
DEBU[0014] Calling POST /v1.24/containers/565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18/start 
DEBU[0014] container mounted via layerStore: /var/lib/docker/aufs/mnt/96bbbd0082d9b4d6bdfe4c362128dd2b16b544d2098a03951edd4cbb8846342e 
DEBU[0014] Assigning addresses for endpoint gigantic_perlman's interface on network bridge 
DEBU[0014] RequestAddress(LocalDefault/172.17.0.0/16, <nil>, map[]) 
DEBU[0014] Assigning addresses for endpoint gigantic_perlman's interface on network bridge 
INFO[0014] No non-localhost DNS nameservers are left in resolv.conf. Using default external servers : [nameserver 8.8.8.8 nameserver 8.8.4.4] 
INFO[0014] IPv6 enabled; Adding default IPv6 external servers : [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844] 
DEBU[0014] Programming external connectivity on endpoint gigantic_perlman (965621cb9405997fc7c079014857f3866b8bf319f4efc21d1ffcfca5ac9e6fe1) 
DEBU[0014] sandbox set key processing took 48.467983ms for container 565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18 
DEBU[0014] libcontainerd: received containerd event: &types.Event{Type:"start-container", Id:"565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18", Status:0x0, Pid:"", Timestamp:(*timestamp.Timestamp)(0xc8208d6020)} 
DEBU[0014] libcontainerd: event unhandled: type:"start-container" id:"565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18" timestamp:<seconds:1471289815 nanos:592861258 >  
DEBU[0014] containerd: process exited                    id=565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18 pid=init status=0 systemPid=29376
DEBU[0014] libcontainerd: received containerd event: &types.Event{Type:"exit", Id:"565cba4a3b718c4a510a6d931a68ab7b7d07183f80450e4eee6beb7ce5f0fc18", Status:0x0, Pid:"init", Timestamp:(*timestamp.Timestamp)(0xc8208b2280)} 
DEBU[0014] attach: stdout: end                          
DEBU[0014] attach: stderr: end                          
```

In the end, it appears this is just another run of the mill inability to detect a process exited issue.  This is related to #1578, #1329, and possibly #858.  We can probably flag this as a duplicate of #1578.

**Comment by sflxn** [comment]
I checked the VIC portlayer logs when running the operation in question.  There is no attempt to attach stdin, which is as expected.  I was just verifying this to be the case.

I think if we add the ability to detect the container exiting in the interaction portlayer, this issue will be resolved... partly.  There is still the fact that docker run calls container wait at the end.

**Comment by sflxn** [comment]
Just chatted with @hickeng.  There is one side effect that we have now that we need to deal with.  The fact that the container is still running even though /bin/sh should have exited.  It appears the tether is creating a stdin for the container regardless of what we specify for attach.  That is why /bin/sh is not exiting and the container continues to run.

We need some mechanism to tell the tether not to create this stdin.  I'm reassigning this to @hickeng as that is more his specialty.

**Comment by hickeng** [comment]
dup of #1444 ?

**Comment by stuclem** [comment]
Already in the release notes, so removing the kind/note tag.

**Comment by mlh78750** [comment]
Release notes already there. There is an easy work around. Need to get resolved before 1.0 but okay to push out of .8.
